### ISSUE TITLE
[master] Race conditions

### ISSUE THREAD

**Issue by cHoco** [koekeishiya/kwm#515]
While trying to understand better how https://github.com/koekeishiya/kwm/issues/511 is happening I decided to run kwm with some sanitizers enabled, in this case, ThreadSanitazer.
I'll report them here as more come by:

1. Data race between `sharedworkspace.mm` and `window.cpp` and more generally almost all notification callbacks in `shareworkspace.mm` In this files we access global variable that are also accessed and modified by the event loop. For example:
The `didActivateApplication:` callback accesses the Applications array and selects an application for which it checks the focus window (line 170), but at the same time the event loop may be processing an AXEvent_WindowFocused in which the focused window is changed (line 465).
Basically, we need to setup mutexes to access these variables, I think some focus border issues were caused by this.
